sematic/ui/packages/common/src/hooks/metricsHooks.ts

```diff
@@ -69,3 +71,67 @@ export default function useBasicMetrics({ runId, rootFunctionPath }: {
 
     return { payload: value, loading, error, totalCount, successRate, avgRuntime };
 }
+
+export type MetricsFilter = {
+    metricName: string;
+    labels: { [k: string]: any };
+    groupBys: string[];
+};
+
+export function useMetrics(
+    metricsFilter: MetricsFilter,
+    broadcastEvent?: MetricPoint[]
+): [MetricsPayload | undefined, boolean, Error | undefined] {
+    const { fetch } = useHttpClient();
+
+    const eventMatchesFilter = useCallback(
+        () =>
+            broadcastEvent &&
+            !!broadcastEvent.find(
+                (metricPoint) =>
+                    metricPoint.name === metricsFilter.metricName &&
+                    [
+                        "run_id",
+                        "function_path",
+                        "root_id",
+                        "root_function_path",
+                    ].every(
+                        (key) =>
+                            // Filter does not filter on key
+                            metricsFilter.labels[key] === undefined ||
+                            // or it does and the value matches
+                            metricsFilter.labels[key] ===
+                                metricPoint.labels[key]
+                    )
+            ),
+        [metricsFilter, broadcastEvent]
+    );
+
+    const { value, loading, error } = useAsync(async () => {
+        if (broadcastEvent && !eventMatchesFilter()) return;
+        const labelsJSON = JSON.stringify(metricsFilter.labels);
+        const groupBysStr = metricsFilter.groupBys.join(",");
+        const response = await fetch({
+            url: `/api/v1/metrics/${metricsFilter.metricName}?labels=${labelsJSON}&group_by=${groupBysStr}&rollup=auto`,
+        });
+        return (await response.json()) as MetricsPayload;
+    }, [broadcastEvent]);
```

Comment:
The dependencies should have `metricsFilter` or `eventMatchesFilter` or there will be a bug if either of these dependencies change. 

Secondly, you can optionally make  `eventMatchesFilter()` stateless by passing `metricsFilter, broadcastEvent` as parameters to it. It can be then extracted outside of the hook function and we avoid a heavy lifting of `useCallback`.

Reply:
Not yet addressed.